src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,449 @@
+                var capturesThis = new HashSet<MethodSymbol>();
+                var capturesVariable = new HashSet<MethodSymbol>();
+                var visitStack = new Stack<MethodSymbol>();
+                VisitClosures(_scopeTree, (scope, closure) =>
+                {
+                    foreach (var capture in closure.CapturedVariables)
+                    {
+                        if (capture is MethodSymbol localFunc)
+                        {
+                            methodGraph.Add(localFunc, closure.Symbol);
+                        }
+                        else if (capture == _topLevelMethod.ThisParameter)
+                        {
+                            if (capturesThis.Add(closure.Symbol))
+                            {
+                                visitStack.Push(closure.Symbol);
+                            }
+                        }
+                        else if (capturesVariable.Add(closure.Symbol) && !capturesThis.Contains(closure.Symbol))
+                        {
+                            visitStack.Push(closure.Symbol);
+                        }
+                    }
+                });
+
+                while (visitStack.Count > 0)
+                {
+                    var current = visitStack.Pop();
+                    var setToAddTo = capturesVariable.Contains(current) ? capturesVariable : capturesThis;
+                    foreach (var capturesCurrent in methodGraph[current])
+                    {
+                        if (setToAddTo.Add(capturesCurrent))
+                        {
+                            visitStack.Push(capturesCurrent);
+                        }
+                    }
+                }
+
+                VisitClosures(_scopeTree, (scope, closure) =>
+                {
+                    if (!capturesVariable.Contains(closure.Symbol))
+                    {
+                        closure.CapturedVariables.Clear();
+                    }
+                    if (capturesThis.Contains(closure.Symbol))
+                    {
+                        closure.CapturedVariables.Add(_topLevelMethod.ThisParameter);
+                    }
+                });
+            }
+
+            /// <summary>
+            /// Visit all closures in all nested scopes and run the <paramref name="action"/>.
+            /// </summary>
+            private static void VisitClosures(Scope scope, Action<Scope, Closure> action)
+            {
+                foreach (var closure in scope.Closures)
+                {
+                    action(scope, closure);
+                }
+                foreach (var nested in scope.NestedScopes)
+                {
+                    VisitClosures(nested, action);
+                }
+            }
+
+            private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+            {
+                private Scope _currentScope;
```

Comment:
In the other types you prefer properties, both readonly and rewrite, here you prefer fields. Is there a reason for the distinction?